progetto/ServerProject/src/server/AccountListMonitor.java

```diff
@@ -1,287 +1,292 @@
+                        return i;
+                    }
+                }
             }
-        }finally{
+        } finally {
             this.length++;
             listLock.writeLock().unlock();
         }
         throw new AccountMonitorRuntimeException("ERRORE:addAccount");//Non dovrebbe mai essere sollevata :D speremmu!
     }
 
-    public Account getAccountCopy(int accountId){
+    public Account addAccount(Account account, int accountId) {
+        testRange(accountId);
 
-        Account snapShot;
-        Account curr;
-
-        if(accountId>=this.MAXACCOUNTNUMBER||accountId<0){
-            throw new IllegalArgumentException("accountId>MAXACCOUNTNUMBER || accountId<0");
-        }
-
-        this.listLock.readLock().lock();
-        try {
-            curr = accountList[accountId];
-            if(curr==null){
-                return null;
-            }
-            snapShot = new
-                    Account(curr.getUsername(), curr.getPassword(), curr.getStub(), curr.getPublicKey(), curr.getAccountId());
-            return snapShot;
-        }
-        finally {
-            listLock.readLock().unlock();
-        }
-    }
-
-    public Account addAccount(Account account,int accountId)
-    {
-        if(accountId>=this.MAXACCOUNTNUMBER||accountId<0){
-            throw new IllegalArgumentException("accountId>MAXACCOUNTNUMBER || accountId<0");
-        }
         Account prev;
         this.listLock.writeLock().lock();
-        try{
-            prev=accountList[accountId];
+        try {
+            prev = accountList[accountId];
 
             //gestione length
-            if(prev!=null&&account==null){
+            if (prev != null && account == null) {
                 this.length--;
-            }else if(prev==null && account!=null){
+            } else if (prev == null && account != null) {
                 this.length++;
             }
-            if(account==null){
-                lastFreeposition=accountId;
-            }else{
+            if (account == null) {
+                lastFreeposition = accountId;
+            } else {
                 account.setAccountId(accountId);
             }
-            accountList[accountId]=account;
-        }finally {
+            accountList[accountId] = account;
+        } finally {
             this.listLock.writeLock().unlock();
         }
         return prev;
     }
 
+    //Secondo me è più veloce avere un metodo specializato che riutilizare addAccount(Account account, int accountId)
+    //Se preferisci come era prima lo si cambia
+    public Account removeAccount(int accountId) {
+        testRange(accountId);
 
-    public Account removeAccount(int accountId)
-    {
-        return this.addAccount(null,accountId);
+        Account toRemove;
+        this.listLock.writeLock().lock();
+        try {
+            toRemove = accountList[accountId];
+            if (toRemove != null) {
+                this.length--;
+                accountList[accountId] = null;
```

Comment:
Bisogna deciderre però come usare la cache qui...
Direi che qui basta mettere in più lastFreeposition=accountId;